libcaf_io/caf/io/basp/instance.hpp

```diff
@@ -320,37 +315,31 @@ class instance {
           if (e)
             return false;
         }
-        // close self connection after handshake is done
+        // Close self connection after handshake is done.
         if (hdr.source_node == this_node_) {
           CAF_LOG_INFO("close connection to self immediately");
           callee_.finalize_handshake(hdr.source_node, aid, sigs);
           return false;
         }
-        // close this connection if we already have a direct connection
-        if (tbl_.lookup_direct(hdr.source_node)) {
+        // Close this connection if we already established communication.
+        // TODO: Should we allow multiple "connections" over different
+        // transport protocols?
+        if (tbl_.lookup(hdr.source_node)) {
           CAF_LOG_INFO("close connection since we already have a "
-                       "direct connection: " << CAF_ARG(hdr.source_node));
+                       "connection: " << CAF_ARG(hdr.source_node));
           callee_.finalize_handshake(hdr.source_node, aid, sigs);
           return false;
         }
-        // add direct route to this node and remove any indirect entry
-        CAF_LOG_INFO("new direct connection:" << CAF_ARG(hdr.source_node));
-        tbl_.add_direct(hdl, hdr.source_node);
-        auto was_indirect = tbl_.erase_indirect(hdr.source_node);
-        // write handshake as client in response
-        auto path = tbl_.lookup(hdr.source_node);
-        if (!path) {
-          CAF_LOG_ERROR("no route to host after server handshake");
-          return false;
-        }
-        if (tcp_based) {
-          auto ch = get<connection_handle>(path->hdl);
-          write_client_handshake(ctx, callee_.get_buffer(ch),
-                                 hdr.source_node);
-        }
-        callee_.learned_new_node_directly(hdr.source_node, was_indirect);
+        // Add this node to our contacts.
+        CAF_LOG_INFO("new endpoint:" << CAF_ARG(hdr.source_node));
+        tbl_.add(hdl, hdr.source_node);
+        // TODO: Add status, addresses, ... ?
```

Comment:
This TODO would add code that wasn't there previously. Why do we need to store additional state now?

Reply:
* *Status:* Track if the node is reachable. Can probably be avoided if we assume that its entry in the table means it is known and an associated handle means it is reachable.
* *Address:* I'd like to buffer the addresses of known nodes to make the propagation easier and avoid asking each time a third node requests them.